### PR TITLE
Data science

### PR DESCRIPTION
**Pull request by jklemen**
Added the tests file. I realize this should have been done beforehand, but I overlooked that line in the exercise. I corrected this in the overlord exercise by writing the tests first, as instructed.

### CODE REVIEW

File: data_science/data_science.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
+require 'json'
+
+class DataScience
+  attr_accessor :cohorts, :sample_size
+
+  def initialize(json_file)
+    json_data = open(json_file).read
```
Comment (jmtai) [inline data_science/data_science.rb]: While this form allows you to concisely read the data from a file, it doesn't allow you to actually close the file after you're done reading from it.  You can use the block form to ensure the file is closed after you're done reading it:
```ruby
open(json_file) { |file| file.read }
```

File: data_science/data_science.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    json = JSON.parse(json_data).to_a
+    @sample_size = json.count
```
Comment (jmtai) [inline data_science/data_science.rb]: Hmmm... this is the total sample size of *all* cohorts. Why is this needed?

File: data_science/data_science.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    parts = json.partition { |item| item['cohort'] == 'A' }
```
Comment (jmtai) [inline data_science/data_science.rb]: What happens if the cohort label is lowercase?

File: data_science/data_science.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    @cohorts = Hash[*%w(A B).zip(parts).flatten(1)]
```
Comment (jmtai) [inline data_science/data_science.rb]: This is certainly concise, but would someone looking at this the first time be able to understand it in 5 seconds?
Do you think something like this is more readable?
```ruby
@cohorts = {
  "A" => parts[0],
  "B" => parts[1]
}
```

File: data_science/data_science.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  def cohort_conversions(cohort_name)
```
Comment (jmtai) [inline data_science/data_science.rb]: Any time you see yourself naming methods with and identical prefix, it can be a hint that maybe these belong in a separate class. In this case, there are several things that are specific to a single cohort (e.g. sample size, conversions, error bars, etc) and some things that aren't (e.g. chi-square test, printing results to the console).

The principle of single responsibility tells us that we'd ideally like our classes to do just one thing (or a group of closely related things), so perhaps it would be better to create a `Cohort` class with a cohort-specific variables and methods?

File: data_science/data_science.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    cohort = @cohorts[cohort_name]
+    cohort.inject(0) { |sum, data| sum + data['result'] }
```
Comment (jmtai) [inline data_science/data_science.rb]: Since you're storing the raw data in `@cohorts`, every time you want to get the number of conversions you have to iterate through the entire array to count the results. Do you ever expect this number to change? If not, wouldn't it be more efficient to simple do this once and store the result so the next time you access it, it's just a simply lookup?

File: data_science/data_science.rb [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  def cohort_stats(cohort_name)
+    conv = cohort_conversions(cohort_name)
+    sample = @cohorts[cohort_name].count
```
Comment (jmtai) [inline data_science/data_science.rb]: Naming can be pretty subjective, but we should definitely try to use the most descriptive names possible for our variables. `conv` and `sample` are pretty terse and somewhat cryptic. What about something like `num_conversions` and `total_samples`?

File: data_science/data_science.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    [conv, sample]
+  end
+
+  # 1. Total sample size and number of conversions for each cohort.
+  def report_sample_sizes
+    @cohorts.keys.each do |name|
+      stats = cohort_stats(name)
+
+      puts "Cohort #{name}: #{stats[0]} / #{stats[1]}"
```
Comment (jmtai) [inline data_science/data_science.rb]: Raw indexed arrays are another indication that maybe there's a better way to represent your data. Ruby is an object oriented language, so why not use that to our advantage? Do you think it would make your code more readable if you could do something like this?
```ruby
@cohorts.each { |cohort| puts "Cohort #{cohort.name}: #{cohort.num_conversions} / #{cohort.total_samples}"
```

File: data_science/data_science.rb [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    end
+  end
+
+  def calc_mean_error(cohort_name)
+    stats = cohort_stats(cohort_name)
+
+    p = (stats[0].to_f / stats[1].to_f)
```
Comment (jmtai) [inline data_science/data_science.rb]: `p` is definitely not a great variable name. What you're calculating here is the conversion rate, so why not just call it that? Also, the Chi-square test calculates a p-value, so one could get a little confused by the use of the name `p` here.

File: data_science/data_science.rb [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    se = (2 * Math.sqrt(p * (1 - p) / @sample_size.to_f))
```
Comment (jmtai) [inline data_science/data_science.rb]: You're calculating the error bars per cohort, but `@sample_size` has been defined to be the size of *all* items in the data set.

File: data_science/data_science.rb [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+      puts "Conversion Rate #{name} = #{p}"
+      puts "Standard Error #{name}  = #{se}"
+    end
+  end
+
+  def calc_confidence(cohort_name)
+    p, se = calc_mean_error(cohort_name)
+
+    [p - se, p, p + se].map { |num| num.round(3) }
```
Comment (jmtai) [inline data_science/data_science.rb]: FYI, the calculate the confidence interval w/95% confidence, you actually need a scaling factor on the standard error value: http://onlinestatbook.com/2/estimation/mean.html

For normal distributions, this ends up being 1.96

File: data_science/data_science.rb [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  # 3. Confidence level that the current leader is in fact better than random.
+  # (done in confidence)
+  def report_confidence
```
Comment (jmtai) [inline data_science/data_science.rb]: This is the confidence interval for the conversion rate, but **not** the confidence level that the current leader is in fact better than random. For that you'll need to use the Chi-square test. You can use the ABAnalyzer gem mentioned in the problem description to do this for you: https://github.com/bmuller/abanalyzer

File: data_science/rspec/test_spec.rb [hunk 13]
```diff
@@ -0,0 +1,54 @@
+require '../data_science.rb'
+
+describe DataScience do
+  before :all do
```
Comment (jmtai) [inline data_science/rspec/test_spec.rb]: `let` is usually preferred here because let blocks are lazily evaluated and also memoized so they are more efficient that `before` blocks.

File: data_science/rspec/test_spec.rb [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    data_file = '../data_export_2014_06_20_15_59_02.json'
+    @ds = DataScience.new(data_file)
+  end
+
+  it "should have the correct sample size" do
+    @ds.sample_size.should == 2892
```
Comment (jmtai) [inline data_science/rspec/test_spec.rb]: The `should` syntax is actually deprecated - you'll want to use the newer `expect` syntax for your tests.

File: data_science/rspec/test_spec.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  it "should have two cohorts" do
+    @ds.cohorts.count.should == 2
+  end
+
+  it "should report the correct cohort sample sizes" do
+    a_stats = @ds.cohort_stats('A')
+    a_stats[1].should == 1349
```
Comment (jmtai) [inline data_science/rspec/test_spec.rb]: Just as we saw in the class implementation, using raw arrays to represent your data makes it less understandable. It'd be nice if the test read more like english - something like
```ruby
expect(cohort.sample_size).to eq(1349)
```

File: data_science/rspec/test_spec.rb [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    (a_stats[0] - 0.028).abs.should <= 0.01
+    (a_stats[1] - 0.035).abs.should <= 0.01
+    (a_stats[2] - 0.042).abs.should <= 0.01
+
+    b_stats = @ds.calc_confidence('B')
+    (b_stats[0] - 0.043).abs.should <= 0.01
+    (b_stats[1] - 0.051).abs.should <= 0.01
+    (b_stats[2] - 0.059).abs.should <= 0.01
+  end
+end
```
Comment (jmtai) [inline data_science/rspec/test_spec.rb]: Once you implement the Chi-square test, you'll need a test for that here, as well.

Comment (jmtai) [inline data_science/rspec/test_spec.rb]: You've got tests for the happy path (i.e. the data file contains no errors). What if the data set is corrupted, though?

File: data_science/data_science.rb (hunk 3 above)
Comment (jklemen) [inline data_science/data_science.rb]: It won't be? My implementation is domain-specific, given two cohorts with names 'A' and 'B'. Should I extend it to accept any general name?

**Comment by jmtai** [comment]
Good start! Try refactoring the code a bit (and add in the missing functionality) and resubmit when you're ready.

File: data_science/data_science.rb (hunk 4 above)
Comment (jklemen) [inline data_science/data_science.rb]: Good point. Definite violation of the 5 second rule.

File: data_science/data_science.rb (hunk 5 above)
Comment (jklemen) [inline data_science/data_science.rb]: I think you're onto something.

File: data_science/data_science.rb (hunk 6 above)
Comment (jklemen) [inline data_science/data_science.rb]: I will be refactoring everything into a 'Cohort' class that will address this problem

File: data_science/data_science.rb (hunk 7 above)
Comment (jklemen) [inline data_science/data_science.rb]: It's difficult to balance descriptive snake-cased names with the 80-character line length rule, but you're right.

File: data_science/data_science.rb (hunk 10 above)
Comment (jklemen) [inline data_science/data_science.rb]: Oh, so I divide by just the sample size of the cohort for the standard error?

File: data_science/data_science.rb (hunk 11 above)
Comment (jklemen) [inline data_science/data_science.rb]: I *believe* this is handled prematurely in calc_mean_error, by mistake. I will propagate the constant down to the correct method.

File: data_science/rspec/test_spec.rb (hunk 14 above)
Comment (jklemen) [inline data_science/rspec/test_spec.rb]: Just learned about this yesterday, will change!

File: data_science/data_science.rb (hunk 10 above)
Comment (jmtai) [inline data_science/data_science.rb]: Correct, that calculation is per-cohort.

File: data_science/data_science.rb (hunk 1 above)
Comment (djkotowski) [inline data_science/data_science.rb]: And therefore can be shortened to.
```ruby
open(json_file, &:read)
```

File: data_science/data_science.rb (hunk 3 above)
Comment (djkotowski) [inline data_science/data_science.rb]: It would be useful to support case-insensitivity.

File: data_science/data_science.rb (hunk 7 above)
Comment (djkotowski) [inline data_science/data_science.rb]: Just be sure not to name your variables anything like `num_conversions_of_cohort_formerly_known_as_a`

File: data_science/data_science.rb (hunk 8 above)
Comment (djkotowski) [inline data_science/data_science.rb]: It may even be ideal to redefine `to_s` on Cohort to provide this functionality.

File: data_science/data_science.rb (hunk 3 above)
Comment (jklemen) [inline data_science/data_science.rb]: noted.

File: data_science/data_science.rb (hunk 8 above)
Comment (jklemen) [inline data_science/data_science.rb]: Great suggestion!

File: data_science/rspec/test_spec.rb (hunk 13 above)
Comment (jklemen) [inline data_science/rspec/test_spec.rb]: Interesting. I will change this.
